core/src/main/java/hudson/util/DirScanner.java

```diff
@@ -130,6 +139,51 @@ public void scan(File dir, FileVisitor visitor) throws IOException {
 
         private static final long serialVersionUID = 1L;
     }
+    
+    private static class DescendantFileSelector implements FileSelector{
+        private Set<String> alreadyDeselected;
+        private FilePath baseDirFP;
+        private int baseDirPathLength;
+
+        private DescendantFileSelector(File basedir){
+            this.baseDirFP = new FilePath(basedir);
+            this.baseDirPathLength = basedir.getPath().length();
+            this.alreadyDeselected = new HashSet<>();
+        }
+        
+        @Override
+        public boolean isSelected(File basedir, String filename, File file) throws BuildException {
+            String parentName = file.getParent();
+            if (parentName.length() > baseDirPathLength) {
+                String parentRelativeName = parentName.substring(baseDirPathLength + File.separator.length());
```

Comment:
Isn't the file separator always length 1? You can access it via `File.separatorChar` which is length 1.